### DISCUSSION TITLE
Getting panics when using SparkFun WM8960 board

### DISCUSSION THREAD

**Discussion by jamdis** [pschatzmann/arduino-audio-tools#1585] [Q&A] [answered]
I'm trying to use the Sparkfun WM8960 board as an ADC.   I've gotten some of the examples from SparkFun's library working. However, I'm getting panics when trying to use it with Audio Tools.  I've tried this using WM8960Stream, and I've also tried using Sparkfun's library to set up the device, and then treating it as a generic i2S input. Both approaches have the same result.

Thanks!

Current version of my code:

`#include "AudioTools.h"
#include "AudioLibs/WM8960Stream.h"

AudioInfo info(44100, 2, 32);
WM8960Stream in; 
CsvOutput<int32_t> out(Serial);
StreamCopy copier(out, in); // copy i2sStream to csvStream

// Arduino Setup
void setup(void) {
    Serial.begin(115200);
    AudioLogger::instance().begin(Serial, AudioLogger::Debug);
    // setup wire 
    Wire.begin(19,21);
    Wire.setClock(10000);   
    auto cfg = in.defaultConfig(RX_MODE);
    cfg.copyFrom(info);
    cfg.i2s_format = I2S_STD_FORMAT; // or try with I2S_LSB_FORMAT
    cfg.wire = &Wire;
    cfg.is_master = false;
    cfg.use_apll = false; 
    //cfg.pin_bck = 14; //default 14
    //cfg.pin_ws = 15; //default 15
    //cfg.pin_data = 22; //default 22
    in.begin(cfg);
    out.begin(info);

}

// Arduino loop - copy data
void loop() {
    copier.copy();`

And the serial output:

`mtb_wm8960_write 0x1 = 0x192
mtb_wm8960_adjust_heaphone_output_volume
_mtb_wm8960_adjust_volume
mtb_wm8960_write 0x28 = 0x58
mtb_wm8960_write 0x29 = 0x158
mtb_wm8960_adjust_heaphone_output_volume
_mtb_wm8960_adjust_volume
mtb_wm8960_write 0x2 = 0x58
mtb_wm8960_write 0x3 = 0x158
mtb_wm8960_activate
mtb_wm8960_write 0x19 = 0xfe
mtb_wm8960_write 0x1a = 0x0
mtb_wm8960_write 0x2f = 0x30
mtb_wm8960_configure_clocking
_mtb_wm8960_setup_pll
mtb_wm8960_write 0x34 = 0x38
mtb_wm8960_write 0x35 = 0x0
mtb_wm8960_write 0x36 = 0x0
mtb_wm8960_write 0x37 = 0x0
mtb_wm8960_write 0x1a = 0x1
mtb_wm8960_write 0x4 = 0x5
mtb_wm8960_write 0x7 = 0x42
Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

Core  1 register dump:
PC      : 0x400df03a  PS      : 0x00060430  A0      : 0x800dfca5  A1      : 0x3ffb1f80  
A2      : 0xa5a5a5a5  A3      : 0x0000009b  A4      : 0x00000001  A5      : 0x00000000  
A6      : 0xb33fffff  A7      : 0xb33fffff  A8      : 0xd5d71b14  A9      : 0x3ffb1f80  
A10     : 0x3ffb900c  A11     : 0x0000003c  A12     : 0x00060020  A13     : 0x3f40a57c  
A14     : 0xb33fffff  A15     : 0xb33fffff  SAR     : 0x0000000f  EXCCAUSE: 0x0000001c  
EXCVADDR: 0xd5d71b14  LBEG    : 0x4008a0c0  LEND    : 0x4008a0cb  LCOUNT  : 0x00000000  

Backtrace: 0x400df037:0x3ffb1f80 0x400dfca2:0x3ffb1fa0 0x400dfff7:0x3ffb1fe0 0x400d3b04:0x3ffb2010 0x400d4262:0x3ffb2070 0x400d4322:0x3ffb2090 0x400d49ac:0x3ffb20f0 0x400d59cd:0x3ffb21a0 0x400dc03b:0x3ffb2290

ELF file SHA256: ee00f9d59829ca5d`

**Comment by pschatzmann** [comment, marked as answer]
You need to decode the Backtrace!

I am not sure cfg.is_master = false; will work. 
I suggest you keep the default values which uses the ESP32 as master

You should also check if the begin() methods returns true (=success)

**Comment by jamdis** [comment]
Thank you very much for your quick reply!  I was using the Arduino IDE version 2, which doesn't allow for any easy way to decode backtraces.  I've switched to PlatformIO to help decode the backtrace, but I've found that the panics don't occur when compiling with PlatformIO.  Not sure why, but happy to be able to move on to the next problem.

**Comment by pschatzmann** [comment]
Here is the link to a decoder that works with version 2:
https://github.com/dankeboy36/esp-exception-decoder

**Reply by jamdis** [reply]
Just wanted to update that I seem to have gotten the Sparkfun board working with audio tools. I'm sticking with PlatformIO and haven't gone back to Arduino IDE. I'm sure I could get backtrace decoding working in Arduino, but PlatformIO has other advantages of course, so I'm not going to mess with it.  Haven't had any of those panics since, but I'll be sure to post here if I do.

In the meantime, the code below (pretty much just  the provided example "streams-12s-serial" combined with Sparkfun's "VolumePlotter" example.) works to get audio into audio_tools from the Sparkfun WM8960 board.   Note that I'm not using the WM8960 library provided by pschatzmann, but rather just using Sparkfun's libarary to set up the codec via i2C and treating the WM8960 as a generic i2S source. I'm curious if anyone has thoughts about the comparative advantages of Sparkfun's library vs. pschatzmann's. 
 
```//*******libraries*********
#include <Arduino.h>
#include <Wire.h>
#include <SparkFun_WM8960_Arduino_Library.h>
#include "AudioTools.h"

//*******pins*********
#define I2S_WS 25 //labelled as ALRC on the board 
#define I2S_SD 17 //labelled as ADAT on the board
#define I2S_SCK 16 //labelled as BCLK on the board

//*******needed for Sparkfun example
// Use I2S Processor 0
#define I2S_PORT I2S_NUM_0

//*********objects*********
//sparkfun
WM8960 codec;
//pschatzman
AudioInfo info(44100, 2, 32);
I2SStream i2sStream; // Access I2S as stream
CsvOutput<int32_t> csvStream(Serial);
StreamCopy copier(csvStream, i2sStream); // copy i2sStream to csvStream

//*********function declarations*********
void setupCodec();

//*********functions*********
void setup() {
  Serial.begin(115200);

  Wire.begin(19,21);
  if (codec.begin() == false) //Begin communication over I2C
  {
    Serial.println("The device did not respond. Please check wiring.");
  }
  else{
    Serial.println("Device is connected properly.");
  }
  setupCodec();

  //setup audioTools' audio logger
  AudioLogger::instance().begin(Serial, AudioLogger::Info);
  auto cfg = i2sStream.defaultConfig(RX_MODE);
    cfg.copyFrom(info);
    cfg.i2s_format = I2S_STD_FORMAT; // or try with I2S_LSB_FORMAT
    cfg.is_master = true; //confirmed does not work if set to false
    cfg.use_apll = false;  // leaving this here, but changing to true seems to have no effect
    cfg.pin_bck = I2S_SCK;
    cfg.pin_data = I2S_SD;
    cfg.pin_ws = I2S_WS;

    i2sStream.begin(cfg);
    csvStream.begin(info);
}

void loop() {
  copier.copy();
}


void setupCodec(){
  // General setup needed
  codec.enableVREF();
  codec.enableVMID();

  // Setup signal flow to the ADC

  codec.enableLMIC();
  codec.enableRMIC();

  // Connect from INPUT1 to "n" (aka inverting) inputs of PGAs.
  codec.connectLMN1();
  codec.connectRMN1();

  // Disable mutes on PGA inputs (aka INTPUT1)
  codec.disableLINMUTE();
  codec.disableRINMUTE();

  // Set pga volumes
  codec.setLINVOLDB(0.00); // Valid options are -17.25dB to +30dB (0.75dB steps)
  codec.setRINVOLDB(0.00); // Valid options are -17.25dB to +30dB (0.75dB steps)

  // Set input boosts to get inputs 1 to the boost mixers
  codec.setLMICBOOST(WM8960_MIC_BOOST_GAIN_0DB);
  codec.setRMICBOOST(WM8960_MIC_BOOST_GAIN_0DB);

  // Connect from MIC inputs (aka pga output) to boost mixers
  codec.connectLMIC2B();
  codec.connectRMIC2B();

  // Enable boost mixers
  codec.enableAINL();
  codec.enableAINR();

  // Connect LB2LO (booster to output mixer (analog bypass)
  codec.enableLB2LO();
  codec.enableRB2RO();

  // Disconnect from DAC outputs to output mixer
  codec.disableLD2LO();
  codec.disableRD2RO();

  // Set gainstage between booster mixer and output mixer
  codec.setLB2LOVOL(WM8960_OUTPUT_MIXER_GAIN_0DB); 
  codec.setRB2ROVOL(WM8960_OUTPUT_MIXER_GAIN_0DB); 

  // Enable output mixers
  codec.enableLOMIX();
  codec.enableROMIX();

  // CLOCK STUFF, These settings will get you 44.1KHz sample rate, and class-d 
  // freq at 705.6kHz
  codec.enablePLL(); // Needed for class-d amp clock
  codec.setPLLPRESCALE(WM8960_PLLPRESCALE_DIV_2);
  codec.setSMD(WM8960_PLL_MODE_FRACTIONAL);
  codec.setCLKSEL(WM8960_CLKSEL_PLL);
  codec.setSYSCLKDIV(WM8960_SYSCLK_DIV_BY_2);
  codec.setBCLKDIV(4);
  codec.setDCLKDIV(WM8960_DCLKDIV_16);
  codec.setPLLN(7);
  codec.setPLLK(0x86, 0xC2, 0x26); // PLLK=86C226h
  //codec.setADCDIV(0); // Default is 000 (what we need for 44.1KHz)
  //codec.setDACDIV(0); // Default is 000 (what we need for 44.1KHz)
  codec.setWL(WM8960_WL_16BIT);

  codec.enablePeripheralMode();
  //codec.enableMasterMode();
  //codec.setALRCGPIO(); // Note, should not be changed while ADC is enabled.

  // Enable ADCs, and disable DACs
  codec.enableAdcLeft();
  codec.enableAdcRight();
  codec.disableDacLeft();
  codec.disableDacRight();
  codec.disableDacMute();

  //codec.enableLoopBack(); // Loopback sends ADC data directly into DAC
  codec.disableLoopBack();

  // Default is "soft mute" on, so we must disable mute to make channels active
  codec.enableDacMute(); 

  codec.enableHeadphones();
  codec.enableOUT3MIX(); // Provides VMID as buffer for headphone ground

  codec.setHeadphoneVolumeDB(0.00);
}
  ```

**Reply by nradina** [reply]
Thank you for this. I too am trying to utilize the Sparkfun WM8960 board (advantages for me are the input options and PGA) with Phil's amazing library.  Is this code example you provided esentailly achieving these two goals?  

Thansk for your help!

**Reply by jamdis** [reply]
@nradina Yeah that's right.  The sparkfun library provides functions to program the registers every which way via i2C so you should be able to set up the PGA any way you like I think, while  audiotools handles the actual audio stream coming in over i2S. Using them together seems to be best of both worlds, but I'm still tinkering with this myself.

**Reply by nradina** [reply]
Wonderful. Thank you! I think this duo would be great! With your code, am i right to assume that the sugnal passes through the PGA and directly to the analog output "mixer", bypassing the ADC -> DAC? I'm tinkering away here..;-> Thanks again!

**Comment by pschatzmann** [comment]
I don't know the Sparkfun libary, so I would not know....
I guess the advantage would be that any audiokit example should theoretically work by just changing the output driver parameter passed to the AudioBoardStream
